rust/chains/tw_ton/src/signer.rs

```diff
@@ -25,35 +22,16 @@ impl TheOpenNetworkSigner {
     }
 
     fn sign_impl(
-        _coin: &dyn CoinContext,
+        coin: &dyn CoinContext,
         input: Proto::SigningInput<'_>,
     ) -> SigningResult<Proto::SigningOutput<'static>> {
-        let signing_request = SigningRequestBuilder::build(&input)?;
-
-        let external_message =
-            ExternalMessageCreator::create_external_message_to_sign(&signing_request)
-                .map_err(cell_to_signing_error)?;
-
-        // Whether to add 'StateInit' reference.
-        let state_init = signing_request.seqno == 0;
-        let signed_tx = signing_request
-            .wallet
-            .sign_transaction(external_message, state_init)
-            .context("Error signing/wrapping an external message")?
-            .build()
-            .context("Error generating signed message cell")
-            .map_err(cell_to_signing_error)?;
+        let msg_to_sign = SigningRequestBuilder::msg_to_sign(&input)?;
+        let wallet = SigningRequestBuilder::build(&input)?.wallet;
```

Comment:
`SigningRequestBuilder::build` is called three times in `TheOpenNetworkSigner::sign_impl`: 
1. `SigningRequestBuilder::msg_to_sign` calls it to generate a message to sign
2. `SigningRequestBuilder::build` itself to get the wallet only.
3. `TheOpenNetworkCompiler::compile_impl` to get the message to be compiled.

`TheOpenNetworkSigner::sign_impl` can be refactored this way:
1. Call `SigningRequestBuilder::build` one time to get both the signing request and wallet.
2. Create an external message via `ExternalMessageCreator::create_external_message_to_sign`.
3. Sign it using `wallet.sign_transaction()`.

`TheOpenNetworkCompiler::pre_image_hash_impl` will do:
1. Call `SigningRequestBuilder::build` one time to get both the signing request.
2. Create an external message via `ExternalMessageCreator::create_external_message_to_sign`.
3. Return the external message hash.

And `TheOpenNetworkCompiler::compile_impl` will do:
1. Call `SigningRequestBuilder::build` one time to get both the signing request.
2. Create an external message via `ExternalMessageCreator::create_external_message_to_sign`.
3. Compile it using `wallet.compile_transaction()`.

Reply:
My initial idea was to have the sign_impl code reuse as much code from other parts as possible to reduce redundancy. Your approach results in some duplicate code between sign_impl and compile_impl, but it's also very good, so I adopted your method.

Reply:
Yes that makes sense of course, but there is a bit duplication that makes the code faster, and uses less memory allocations